File: Problems 11-20/Euler15.hs
-- See Euler15.cpp for an explanation of the solution

main = print $ 40 `choose` 20

choose :: (Integral a) => a -> a -> a
choose n k = round $ foldl f 1.0 [1..k]
    where n' = fromIntegral n
          f = (\a i -> let i' = fromIntegral i in a * (n' + 1 - i') / i')

File: Problems 11-20/Euler15.py
# See Euler15.cpp for an explanation of the solution

def factorial(n):
    if n == 1:
        return 1
    else:
        return n * factorial(n - 1)

def choose(n, k):
    return factorial(n) // (factorial(k) * factorial(n - k))

routes = choose(40, 20)
print(routes)
